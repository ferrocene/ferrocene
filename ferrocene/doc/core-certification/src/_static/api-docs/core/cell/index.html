<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Shareable mutable containers."><title>core::cell - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-916cea96.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="core" data-themes="" data-resource-suffix="1.88.0" data-rustdoc-version="1.87.0 (17067e9ac 2025-05-09)" data-channel="1.87.0" data-search-js="search-e7298875.js" data-settings-js="settings-d72f25bb.js" ><script src="../../static.files/storage-82c7156e.js"></script><script defer src="../sidebar-items1.88.0.js"></script><script defer src="../../static.files/main-fb8c74a8.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button><a class="logo-container" href="../../core/index.html"><img class="rust-logo" src="../../static.files/rust-logo-9a9549ea.svg" alt=""></a></nav><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../../core/index.html"><img class="rust-logo" src="../../static.files/rust-logo-9a9549ea.svg" alt="logo"></a><h2><a href="../../core/index.html">core</a><span class="version">1.88.0</span></h2></div><div class="version">(e925a16bd	2025-08-18)</div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module cell</a></h2><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#cellt" title="`Cell&#60;T&#62;`"><code>Cell&lt;T&gt;</code></a></li><li><a href="#refcellt" title="`RefCell&#60;T&#62;`"><code>RefCell&lt;T&gt;</code></a></li><li><a href="#oncecellt" title="`OnceCell&#60;T&#62;`"><code>OnceCell&lt;T&gt;</code></a></li><li><a href="#lazycellt-f" title="`LazyCell&#60;T, F&#62;`"><code>LazyCell&lt;T, F&gt;</code></a></li><li><a href="#when-to-choose-interior-mutability" title="When to choose interior mutability">When to choose interior mutability</a><ul><li><a href="#introducing-mutability-inside-of-something-immutable" title="Introducing mutability ‘inside’ of something immutable">Introducing mutability ‘inside’ of something immutable</a></li><li><a href="#implementation-details-of-logically-immutable-methods" title="Implementation details of logically-immutable methods">Implementation details of logically-immutable methods</a></li><li><a href="#mutating-implementations-of-clone" title="Mutating implementations of `Clone`">Mutating implementations of <code>Clone</code></a></li></ul></li></ul><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate core</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">core</a></div><h1>Module <span>cell</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><span class="since" title="Stable since Rust version 1.6.0">1.6.0</span> · <a class="src" href="../../src/core/cell.rs.html#1-2586">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Shareable mutable containers.</p>
<p>Rust memory safety is based on this rule: Given an object <code>T</code>, it is only possible to
have one of the following:</p>
<ul>
<li>Several immutable references (<code>&amp;T</code>) to the object (also known as <strong>aliasing</strong>).</li>
<li>One mutable reference (<code>&amp;mut T</code>) to the object (also known as <strong>mutability</strong>).</li>
</ul>
<p>This is enforced by the Rust compiler. However, there are situations where this rule is not
flexible enough. Sometimes it is required to have multiple references to an object and yet
mutate it.</p>
<p>Shareable mutable containers exist to permit mutability in a controlled manner, even in the
presence of aliasing. [<code>Cell&lt;T&gt;</code>], [<code>RefCell&lt;T&gt;</code>], and [<code>OnceCell&lt;T&gt;</code>] allow doing this in
a single-threaded way—they do not implement <a href="../../std/marker/trait.Sync.html"><code>Sync</code></a>. (If you need to do aliasing and
mutation among multiple threads, <a href="../../std/sync/struct.Mutex.html"><code>Mutex&lt;T&gt;</code></a>, <a href="../../std/sync/struct.RwLock.html"><code>RwLock&lt;T&gt;</code></a>, <a href="../../std/sync/struct.OnceLock.html"><code>OnceLock&lt;T&gt;</code></a> or <a href="../sync/atomic/index.html" title="mod core::sync::atomic"><code>atomic</code></a>
types are the correct data structures to do so).</p>
<p>Values of the <code>Cell&lt;T&gt;</code>, <code>RefCell&lt;T&gt;</code>, and <code>OnceCell&lt;T&gt;</code> types may be mutated through shared
references (i.e. the common <code>&amp;T</code> type), whereas most Rust types can only be mutated through
unique (<code>&amp;mut T</code>) references. We say these cell types provide ‘interior mutability’
(mutable via <code>&amp;T</code>), in contrast with typical Rust types that exhibit ‘inherited mutability’
(mutable only via <code>&amp;mut T</code>).</p>
<p>Cell types come in four flavors: <code>Cell&lt;T&gt;</code>, <code>RefCell&lt;T&gt;</code>, <code>OnceCell&lt;T&gt;</code>, and <code>LazyCell&lt;T&gt;</code>.
Each provides a different way of providing safe interior mutability.</p>
<h3 id="cellt"><a class="doc-anchor" href="#cellt">§</a><code>Cell&lt;T&gt;</code></h3>
<p>[<code>Cell&lt;T&gt;</code>] implements interior mutability by moving values in and out of the cell. That is, an
<code>&amp;mut T</code> to the inner value can never be obtained, and the value itself cannot be directly
obtained without replacing it with something else. Both of these rules ensure that there is
never more than one reference pointing to the inner value. This type provides the following
methods:</p>
<ul>
<li>For types that implement <a href="../marker/trait.Copy.html" title="trait core::marker::Copy"><code>Copy</code></a>, the <a href="Cell::get"><code>get</code></a> method retrieves the current
interior value by duplicating it.</li>
<li>For types that implement <a href="../default/trait.Default.html" title="trait core::default::Default"><code>Default</code></a>, the <a href="Cell::take"><code>take</code></a> method replaces the current
interior value with <a href="../default/trait.Default.html#tymethod.default" title="associated function core::default::Default::default"><code>Default::default()</code></a> and returns the replaced value.</li>
<li>All types have:
<ul>
<li><a href="Cell::replace"><code>replace</code></a>: replaces the current interior value and returns the replaced
value.</li>
<li><a href="Cell::into_inner"><code>into_inner</code></a>: this method consumes the <code>Cell&lt;T&gt;</code> and returns the
interior value.</li>
<li><a href="Cell::set"><code>set</code></a>: this method replaces the interior value, dropping the replaced value.</li>
</ul>
</li>
</ul>
<p><code>Cell&lt;T&gt;</code> is typically used for more simple types where copying or moving values isn’t too
resource intensive (e.g. numbers), and should usually be preferred over other cell types when
possible. For larger and non-copy types, <code>RefCell</code> provides some advantages.</p>
<h3 id="refcellt"><a class="doc-anchor" href="#refcellt">§</a><code>RefCell&lt;T&gt;</code></h3>
<p>[<code>RefCell&lt;T&gt;</code>] uses Rust’s lifetimes to implement “dynamic borrowing”, a process whereby one can
claim temporary, exclusive, mutable access to the inner value. Borrows for <code>RefCell&lt;T&gt;</code>s are
tracked at <em>runtime</em>, unlike Rust’s native reference types which are entirely tracked
statically, at compile time.</p>
<p>An immutable reference to a <code>RefCell</code>’s inner value (<code>&amp;T</code>) can be obtained with
<a href="%60RefCell::borrow%60"><code>borrow</code></a>, and a mutable borrow (<code>&amp;mut T</code>) can be obtained with
<a href="%60RefCell::borrow_mut%60"><code>borrow_mut</code></a>. When these functions are called, they first verify that
Rust’s borrow rules will be satisfied: any number of immutable borrows are allowed or a
single mutable borrow is allowed, but never both. If a borrow is attempted that would violate
these rules, the thread will panic.</p>
<p>The corresponding <a href="../../std/marker/trait.Sync.html"><code>Sync</code></a> version of <code>RefCell&lt;T&gt;</code> is <a href="../../std/sync/struct.RwLock.html"><code>RwLock&lt;T&gt;</code></a>.</p>
<h3 id="oncecellt"><a class="doc-anchor" href="#oncecellt">§</a><code>OnceCell&lt;T&gt;</code></h3>
<p>[<code>OnceCell&lt;T&gt;</code>] is somewhat of a hybrid of <code>Cell</code> and <code>RefCell</code> that works for values that
typically only need to be set once. This means that a reference <code>&amp;T</code> can be obtained without
moving or copying the inner value (unlike <code>Cell</code>) but also without runtime checks (unlike
<code>RefCell</code>). However, its value can also not be updated once set unless you have a mutable
reference to the <code>OnceCell</code>.</p>
<p><code>OnceCell</code> provides the following methods:</p>
<ul>
<li><a href="OnceCell::get"><code>get</code></a>: obtain a reference to the inner value</li>
<li><a href="OnceCell::set"><code>set</code></a>: set the inner value if it is unset (returns a <code>Result</code>)</li>
<li><a href="OnceCell::get_or_init"><code>get_or_init</code></a>: return the inner value, initializing it if needed</li>
<li><a href="OnceCell::get_mut"><code>get_mut</code></a>: provide a mutable reference to the inner value, only available
if you have a mutable reference to the cell itself.</li>
</ul>
<p>The corresponding <a href="../../std/marker/trait.Sync.html"><code>Sync</code></a> version of <code>OnceCell&lt;T&gt;</code> is <a href="../../std/sync/struct.OnceLock.html"><code>OnceLock&lt;T&gt;</code></a>.</p>
<h3 id="lazycellt-f"><a class="doc-anchor" href="#lazycellt-f">§</a><code>LazyCell&lt;T, F&gt;</code></h3>
<p>A common pattern with OnceCell is, for a given OnceCell, to use the same function on every
call to [<code>OnceCell::get_or_init</code>] with that cell. This is what is offered by [<code>LazyCell</code>],
which pairs cells of <code>T</code> with functions of <code>F</code>, and always calls <code>F</code> before it yields <code>&amp;T</code>.
This happens implicitly by simply attempting to dereference the LazyCell to get its contents,
so its use is much more transparent with a place which has been initialized by a constant.</p>
<p>More complicated patterns that don’t fit this description can be built on <code>OnceCell&lt;T&gt;</code> instead.</p>
<p><code>LazyCell</code> works by providing an implementation of <code>impl Deref</code> that calls the function,
so you can just use it by dereference (e.g. <code>*lazy_cell</code> or <code>lazy_cell.deref()</code>).</p>
<p>The corresponding <a href="../../std/marker/trait.Sync.html"><code>Sync</code></a> version of <code>LazyCell&lt;T, F&gt;</code> is <a href="../../std/sync/struct.LazyLock.html"><code>LazyLock&lt;T, F&gt;</code></a>.</p>
<h2 id="when-to-choose-interior-mutability"><a class="doc-anchor" href="#when-to-choose-interior-mutability">§</a>When to choose interior mutability</h2>
<p>The more common inherited mutability, where one must have unique access to mutate a value, is
one of the key language elements that enables Rust to reason strongly about pointer aliasing,
statically preventing crash bugs. Because of that, inherited mutability is preferred, and
interior mutability is something of a last resort. Since cell types enable mutation where it
would otherwise be disallowed though, there are occasions when interior mutability might be
appropriate, or even <em>must</em> be used, e.g.</p>
<ul>
<li>Introducing mutability ‘inside’ of something immutable</li>
<li>Implementation details of logically-immutable methods.</li>
<li>Mutating implementations of <a href="../clone/trait.Clone.html" title="trait core::clone::Clone"><code>Clone</code></a>.</li>
</ul>
<h3 id="introducing-mutability-inside-of-something-immutable"><a class="doc-anchor" href="#introducing-mutability-inside-of-something-immutable">§</a>Introducing mutability ‘inside’ of something immutable</h3>
<p>Many shared smart pointer types, including <a href="../../std/rc/struct.Rc.html"><code>Rc&lt;T&gt;</code></a> and <a href="../../std/sync/struct.Arc.html"><code>Arc&lt;T&gt;</code></a>, provide containers that can
be cloned and shared between multiple parties. Because the contained values may be
multiply-aliased, they can only be borrowed with <code>&amp;</code>, not <code>&amp;mut</code>. Without cells it would be
impossible to mutate data inside of these smart pointers at all.</p>
<p>It’s very common then to put a <code>RefCell&lt;T&gt;</code> inside shared pointer types to reintroduce
mutability:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>std::cell::{RefCell, RefMut};
<span class="kw">use </span>std::collections::HashMap;
<span class="kw">use </span>std::rc::Rc;

<span class="kw">fn </span>main() {
    <span class="kw">let </span>shared_map: Rc&lt;RefCell&lt;<span class="kw">_</span>&gt;&gt; = Rc::new(RefCell::new(HashMap::new()));
    <span class="comment">// Create a new block to limit the scope of the dynamic borrow
    </span>{
        <span class="kw">let </span><span class="kw-2">mut </span>map: RefMut&lt;<span class="lifetime">'_</span>, <span class="kw">_</span>&gt; = shared_map.borrow_mut();
        map.insert(<span class="string">"africa"</span>, <span class="number">92388</span>);
        map.insert(<span class="string">"kyoto"</span>, <span class="number">11837</span>);
        map.insert(<span class="string">"piccadilly"</span>, <span class="number">11826</span>);
        map.insert(<span class="string">"marbles"</span>, <span class="number">38</span>);
    }

    <span class="comment">// Note that if we had not let the previous borrow of the cache fall out
    // of scope then the subsequent borrow would cause a dynamic thread panic.
    // This is the major hazard of using `RefCell`.
    </span><span class="kw">let </span>total: i32 = shared_map.borrow().values().sum();
    <span class="macro">println!</span>(<span class="string">"{total}"</span>);
}</code></pre><a class="test-arrow" target="_blank" title="Run code" href="https://play.rust-lang.org/?code=%23!%5Ballow(unused)%5D%0Ause+std::cell::%7BRefCell,+RefMut%7D;%0Ause+std::collections::HashMap;%0Ause+std::rc::Rc;%0A%0Afn+main()+%7B%0A++++let+shared_map:+Rc%3CRefCell%3C_%3E%3E+=+Rc::new(RefCell::new(HashMap::new()));%0A++++//+Create+a+new+block+to+limit+the+scope+of+the+dynamic+borrow%0A++++%7B%0A++++++++let+mut+map:+RefMut%3C'_,+_%3E+=+shared_map.borrow_mut();%0A++++++++map.insert(%22africa%22,+92388);%0A++++++++map.insert(%22kyoto%22,+11837);%0A++++++++map.insert(%22piccadilly%22,+11826);%0A++++++++map.insert(%22marbles%22,+38);%0A++++%7D%0A%0A++++//+Note+that+if+we+had+not+let+the+previous+borrow+of+the+cache+fall+out%0A++++//+of+scope+then+the+subsequent+borrow+would+cause+a+dynamic+thread+panic.%0A++++//+This+is+the+major+hazard+of+using+%60RefCell%60.%0A++++let+total:+i32+=+shared_map.borrow().values().sum();%0A++++println!(%22%7Btotal%7D%22);%0A%7D&amp;edition=2024"></a></div>
<p>Note that this example uses <code>Rc&lt;T&gt;</code> and not <code>Arc&lt;T&gt;</code>. <code>RefCell&lt;T&gt;</code>s are for single-threaded
scenarios. Consider using <a href="../../std/sync/struct.RwLock.html"><code>RwLock&lt;T&gt;</code></a> or <a href="../../std/sync/struct.Mutex.html"><code>Mutex&lt;T&gt;</code></a> if you need shared mutability in a
multi-threaded situation.</p>
<h3 id="implementation-details-of-logically-immutable-methods"><a class="doc-anchor" href="#implementation-details-of-logically-immutable-methods">§</a>Implementation details of logically-immutable methods</h3>
<p>Occasionally it may be desirable not to expose in an API that there is mutation happening
“under the hood”. This may be because logically the operation is immutable, but e.g., caching
forces the implementation to perform mutation; or because you must employ mutation to implement
a trait method that was originally defined to take <code>&amp;self</code>.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>std::cell::OnceCell;

<span class="kw">struct </span>Graph {
    edges: Vec&lt;(i32, i32)&gt;,
    span_tree_cache: OnceCell&lt;Vec&lt;(i32, i32)&gt;&gt;
}

<span class="kw">impl </span>Graph {
    <span class="kw">fn </span>minimum_spanning_tree(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; Vec&lt;(i32, i32)&gt; {
        <span class="self">self</span>.span_tree_cache
            .get_or_init(|| <span class="self">self</span>.calc_span_tree())
            .clone()
    }

    <span class="kw">fn </span>calc_span_tree(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; Vec&lt;(i32, i32)&gt; {
        <span class="comment">// Expensive computation goes here
        </span><span class="macro">vec!</span>[]
    }
}</code></pre><a class="test-arrow" target="_blank" title="Run code" href="https://play.rust-lang.org/?code=%23!%5Ballow(unused)%5D%0A%23!%5Ballow(dead_code)%5D%0A%0Afn+main()+%7B%0A++++use+std::cell::OnceCell;%0A++++%0A++++struct+Graph+%7B%0A++++++++edges:+Vec%3C(i32,+i32)%3E,%0A++++++++span_tree_cache:+OnceCell%3CVec%3C(i32,+i32)%3E%3E%0A++++%7D%0A++++%0A++++impl+Graph+%7B%0A++++++++fn+minimum_spanning_tree(%26self)+-%3E+Vec%3C(i32,+i32)%3E+%7B%0A++++++++++++self.span_tree_cache%0A++++++++++++++++.get_or_init(%7C%7C+self.calc_span_tree())%0A++++++++++++++++.clone()%0A++++++++%7D%0A++++%0A++++++++fn+calc_span_tree(%26self)+-%3E+Vec%3C(i32,+i32)%3E+%7B%0A++++++++++++//+Expensive+computation+goes+here%0A++++++++++++vec!%5B%5D%0A++++++++%7D%0A++++%7D%0A%7D&amp;edition=2024"></a></div>
<h3 id="mutating-implementations-of-clone"><a class="doc-anchor" href="#mutating-implementations-of-clone">§</a>Mutating implementations of <code>Clone</code></h3>
<p>This is simply a special - but common - case of the previous: hiding mutability for operations
that appear to be immutable. The <a href="../clone/trait.Clone.html#tymethod.clone" title="method core::clone::Clone::clone"><code>clone</code></a> method is expected to not change the
source value, and is declared to take <code>&amp;self</code>, not <code>&amp;mut self</code>. Therefore, any mutation that
happens in the <code>clone</code> method must use cell types. For example, <a href="../../std/rc/struct.Rc.html"><code>Rc&lt;T&gt;</code></a> maintains its
reference counts within a <code>Cell&lt;T&gt;</code>.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>std::cell::Cell;
<span class="kw">use </span>std::ptr::NonNull;
<span class="kw">use </span>std::process::abort;
<span class="kw">use </span>std::marker::PhantomData;

<span class="kw">struct </span>Rc&lt;T: <span class="question-mark">?</span>Sized&gt; {
    ptr: NonNull&lt;RcInner&lt;T&gt;&gt;,
    phantom: PhantomData&lt;RcInner&lt;T&gt;&gt;,
}

<span class="kw">struct </span>RcInner&lt;T: <span class="question-mark">?</span>Sized&gt; {
    strong: Cell&lt;usize&gt;,
    refcount: Cell&lt;usize&gt;,
    value: T,
}

<span class="kw">impl</span>&lt;T: <span class="question-mark">?</span>Sized&gt; Clone <span class="kw">for </span>Rc&lt;T&gt; {
    <span class="kw">fn </span>clone(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; Rc&lt;T&gt; {
        <span class="self">self</span>.inc_strong();
        Rc {
            ptr: <span class="self">self</span>.ptr,
            phantom: PhantomData,
        }
    }
}

<span class="kw">trait </span>RcInnerPtr&lt;T: <span class="question-mark">?</span>Sized&gt; {

    <span class="kw">fn </span>inner(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; <span class="kw-2">&amp;</span>RcInner&lt;T&gt;;

    <span class="kw">fn </span>strong(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; usize {
        <span class="self">self</span>.inner().strong.get()
    }

    <span class="kw">fn </span>inc_strong(<span class="kw-2">&amp;</span><span class="self">self</span>) {
        <span class="self">self</span>.inner()
            .strong
            .set(<span class="self">self</span>.strong()
                     .checked_add(<span class="number">1</span>)
                     .unwrap_or_else(|| abort() ));
    }
}

<span class="kw">impl</span>&lt;T: <span class="question-mark">?</span>Sized&gt; RcInnerPtr&lt;T&gt; <span class="kw">for </span>Rc&lt;T&gt; {
   <span class="kw">fn </span>inner(<span class="kw-2">&amp;</span><span class="self">self</span>) -&gt; <span class="kw-2">&amp;</span>RcInner&lt;T&gt; {
       <span class="kw">unsafe </span>{
           <span class="self">self</span>.ptr.as_ref()
       }
   }
}</code></pre><a class="test-arrow" target="_blank" title="Run code" href="https://play.rust-lang.org/?code=%23!%5Ballow(unused)%5D%0Afn+main()+%7B%0A++++use+std::cell::Cell;%0A++++use+std::ptr::NonNull;%0A++++use+std::process::abort;%0A++++use+std::marker::PhantomData;%0A++++%0A++++struct+Rc%3CT:+?Sized%3E+%7B%0A++++++++ptr:+NonNull%3CRcInner%3CT%3E%3E,%0A++++++++phantom:+PhantomData%3CRcInner%3CT%3E%3E,%0A++++%7D%0A++++%0A++++struct+RcInner%3CT:+?Sized%3E+%7B%0A++++++++strong:+Cell%3Cusize%3E,%0A++++++++refcount:+Cell%3Cusize%3E,%0A++++++++value:+T,%0A++++%7D%0A++++%0A++++impl%3CT:+?Sized%3E+Clone+for+Rc%3CT%3E+%7B%0A++++++++fn+clone(%26self)+-%3E+Rc%3CT%3E+%7B%0A++++++++++++self.inc_strong();%0A++++++++++++Rc+%7B%0A++++++++++++++++ptr:+self.ptr,%0A++++++++++++++++phantom:+PhantomData,%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A++++%0A++++trait+RcInnerPtr%3CT:+?Sized%3E+%7B%0A++++%0A++++++++fn+inner(%26self)+-%3E+%26RcInner%3CT%3E;%0A++++%0A++++++++fn+strong(%26self)+-%3E+usize+%7B%0A++++++++++++self.inner().strong.get()%0A++++++++%7D%0A++++%0A++++++++fn+inc_strong(%26self)+%7B%0A++++++++++++self.inner()%0A++++++++++++++++.strong%0A++++++++++++++++.set(self.strong()%0A+++++++++++++++++++++++++.checked_add(1)%0A+++++++++++++++++++++++++.unwrap_or_else(%7C%7C+abort()+));%0A++++++++%7D%0A++++%7D%0A++++%0A++++impl%3CT:+?Sized%3E+RcInnerPtr%3CT%3E+for+Rc%3CT%3E+%7B%0A+++++++fn+inner(%26self)+-%3E+%26RcInner%3CT%3E+%7B%0A+++++++++++unsafe+%7B%0A+++++++++++++++self.ptr.as_ref()%0A+++++++++++%7D%0A+++++++%7D%0A++++%7D%0A%7D&amp;edition=2024"></a></div>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.UnsafeCell.html" title="struct core::cell::UnsafeCell">Unsafe<wbr>Cell</a></dt><dd>The core primitive for interior mutability in Rust.</dd></dl></section></div></main></body></html>