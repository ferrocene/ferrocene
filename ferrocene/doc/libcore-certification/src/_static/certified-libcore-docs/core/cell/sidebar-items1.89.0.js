window.SIDEBAR_ITEMS = {"struct":["UnsafeCell"]};