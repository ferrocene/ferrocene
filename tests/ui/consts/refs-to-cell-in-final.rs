#![feature(const_refs_to_cell)]

use std::cell::*;

struct SyncPtr<T> { x : *const T }
unsafe impl<T> Sync for SyncPtr<T> {}

// These pass the lifetime checks because of the "tail expression" / "outer scope" rule.
// (This relies on `SyncPtr` being a curly brace struct.)
// However, we intern the inner memory as read-only.
// The resulting constant would pass all validation checks, so it is crucial that this gets rejected
// by static const checks!
static RAW_SYNC_S: SyncPtr<Cell<i32>> = SyncPtr { x: &Cell::new(42) };
//~^ ERROR: cannot refer to interior mutable data
const RAW_SYNC_C: SyncPtr<Cell<i32>> = SyncPtr { x: &Cell::new(42) };
//~^ ERROR: cannot refer to interior mutable data

// This one does not get promoted because of `Drop`, and then enters interesting codepaths because
// as a value it has no interior mutability, but as a type it does. See
// <https://github.com/rust-lang/rust/issues/121610>. Value-based reasoning for interior mutability
// is questionable (https://github.com/rust-lang/unsafe-code-guidelines/issues/493) so for now we
// reject this, though not with a great error message.
pub enum JsValue {
    Undefined,
    Object(Cell<bool>),
}
impl Drop for JsValue {
    fn drop(&mut self) {}
}
const UNDEFINED: &JsValue = &JsValue::Undefined;
//~^ WARNING: mutable pointer in final value of constant
//~| WARNING: this was previously accepted by the compiler but is being phased out

// In contrast, this one works since it is being promoted.
const NONE: &'static Option<Cell<i32>> = &None;
// Making it clear that this is promotion, not "outer scope".
const NONE_EXPLICIT_PROMOTED: &'static Option<Cell<i32>> = {
    let x = &None;
    x
};

fn main() {}
