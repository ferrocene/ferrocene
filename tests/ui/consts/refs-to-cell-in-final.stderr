error[E0492]: statics cannot refer to interior mutable data
  --> $DIR/refs-to-cell-in-final.rs:13:54
   |
LL | static RAW_SYNC_S: SyncPtr<Cell<i32>> = SyncPtr { x: &Cell::new(42) };
   |                                                      ^^^^^^^^^^^^^^ this borrow of an interior mutable value may end up in the final value
   |
   = help: to fix this, the value can be extracted to a separate `static` item and then referenced

error[E0492]: constants cannot refer to interior mutable data
  --> $DIR/refs-to-cell-in-final.rs:15:53
   |
LL | const RAW_SYNC_C: SyncPtr<Cell<i32>> = SyncPtr { x: &Cell::new(42) };
   |                                                     ^^^^^^^^^^^^^^ this borrow of an interior mutable value may end up in the final value

warning: encountered mutable pointer in final value of constant
  --> $DIR/refs-to-cell-in-final.rs:30:1
   |
LL | const UNDEFINED: &JsValue = &JsValue::Undefined;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #122153 <https://github.com/rust-lang/rust/issues/122153>
   = note: `#[warn(const_eval_mutable_ptr_in_final_value)]` on by default

error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0492`.
Future incompatibility report: Future breakage diagnostic:
warning: encountered mutable pointer in final value of constant
  --> $DIR/refs-to-cell-in-final.rs:30:1
   |
LL | const UNDEFINED: &JsValue = &JsValue::Undefined;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #122153 <https://github.com/rust-lang/rust/issues/122153>
   = note: `#[warn(const_eval_mutable_ptr_in_final_value)]` on by default

